internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+			// Existing identity certificate is valid.
+			return nil
+		}
+	}
+
+	// If we get here a new identity certificate must be created.
+	m.idCert, err = createIDCert(m.signer)
+	if err != nil {
+		return err
+	}
+	err = saveIDCert(idPath, m.idCert)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// loadIDCert loads the x509 identity certificate, which contains
+// the identity public key. We only allow a single certificate
+// block in the PEM file.
+func loadIDCert(path string) (*x509.Certificate, error) {
+	err := expectPermission(path, 0o600)
+	if err != nil {
+		return nil, err
+	}
+
+	var cert *x509.Certificate
+	pemData, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	blockCount := 0
+	for {
+		var block *pem.Block
+		block, pemData = pem.Decode(pemData)
+		if block == nil {
+			if blockCount == 1 {
+				break
+			}
+			return nil, fmt.Errorf("empty PEM file")
+		}
+		switch block.Type {
+		case "CERTIFICATE":
+			// We only support a single certificate block.
+			if blockCount == 1 {
+				return nil, fmt.Errorf("unexpected PEM block")
+			}
+			cert, err = x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, err
+			}
+			blockCount += 1
+		default:
+			return nil, fmt.Errorf("unexpected PEM block")
+		}
+	}
+	return cert, nil
+}
+
+// saveIDCert saves the x509 identity certificate, which contains
+// the identity public key.
+func saveIDCert(path string, cert *x509.Certificate) error {
+	certPEMBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert.Raw,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, certPEMBlock); err != nil {
+		return err
+	}
+	pemBytes := pemBuffer.Bytes()
+
+	pemFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+
+	if _, err = pemFile.Write(pemBytes); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createIDCert create a public x509 CA certificate, signed by the identity
+// key. This certificate is included in the TLS certificate chain as the
+// non-leaf certificate, allowing the client to pin this certificate during
+// the client server trust exchange (pairing) procedure.
+func createIDCert(signer crypto.Signer) (*x509.Certificate, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, err
+	}
+	template := &x509.Certificate{
+		SerialNumber:          serialNumber,
+		NotBefore:             systemTime(),
+		NotAfter:              systemTime().Add(idCertValidity),
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		// We can only sign leaf certificates with this.
+		MaxPathLen:     0,
+		MaxPathLenZero: true,
+	}
```

Comment:
The subject field seems to be missing. I am not sure what go does in this case, but according to the RFC, the subject field is required: https://www.rfc-editor.org/rfc/rfc5280#section-4.1.2.6

In this case, it might be unnecessary if the client and server are from the same code base, but if it is expected to have other clients, this might break validation. I would suggest populating it at least with something like a uuid.

Reply:
@ghislainbourgeois 

I have now addressed this. By default, I now generate at least a Subject Common Name (which Go appears to copy to the Issuer Common Name if the Issuer is left empty.

The Common Name now looks like this:

```<Program Name>-<Fingerprint[:8]>```

which means for Pebble it will look something like this:

```pebble-A453BD6C```

For derivative, the program name will change. But also, there is a ```SetX509Templates``` method now added that will allow templates to be provided externally, which can be used to supply ```Subject```, ```Issuer``` and the ```SAN``` fields (```DNSNames``` and ```EmailAddresses```) for now.

I think this should be sufficient for now. What do you think ?

Reply:
Resolving this comment.